Automattic/wp-calypso
# AB Test for Business bump offer

#### Changes proposed in this Pull Request

* This PR implements the test plan explained here pbxNRc-lW-p2

#### Testing instructions

**Scenario 1** _(Testing behavior of `showBusinessPlanBump` A/B test)_

* Signup for a new account from http://calypso.localhost:3000/start.
* Add a Premium plan to cart and complete the purchase
* Verify the following:
    - That you are shown the Business upgrade upsell if on the `variantShowPlanBump` variation of `showBusinessPlanBump` test.
    - That you are shown the concierge upsell if on the `control` variation of `showBusinessPlanBump` test (no matter if you're in the `offer` or `noOffer`group for the `conciergeUpsellDial` test).

If on the plan upsell nudge, verify the following:
- Clicking the `No thanks ...` button should take you to the checklist page.
- Clicking the `Yes, I'd love to try ....` button should take you to the checkout page with the Business plan added to cart and **credits applied**. Verify that after you complete the purchase, you are **not shown the concierge upsell** but instead taken to the checklist/customer home page.

If on the concierge upsell nudge, verify the following:
- Clicking Skip on the nudge should take you to the checklist.
- Clicking Accept on the nudge, and completing the concierge session purchase should take you the checklist. _**No more upsell nudges should be shown**_.

**Scenario 2** _(Test that an invalid card(checkout error) does not lead to a test assignment)_
* Go through the signup flow and checkout with a Premium plan. 
* In the Checkout page, use an invalid visa card number and click the Purchase button(I used `4539747991309129 `).
* You should get a Transaction declined message.
* In your browser console, type `localStorage.ABTests;`. The object should not include the `showBusinessPlanBump` test.
* Now complete the purchase with a valid credit card(which can just be the store sandbox credit card number).
* In your browser console, type `localStorage.ABTests;`. The object _should include_ the `showBusinessPlanBump` test.

**Scenario 3** _(Test that the offer is shown when upgrading from a lower-tier plan to Premium)_
* Login to a site having a Free or Blogger or Personal plan, go to /plans page and upgrade to Premium
* After purchase completes, verify the following:
    - That you are shown the Business upgrade upsell if on the `variantShowPlanBump` variation of `showBusinessPlanBump` test.
    - That you are shown the concierge upsell if on the `control` variation of `showBusinessPlanBump` test (no matter if you're in the `offer` or `noOffer` group for the `conciergeUpsellDial` test).
In both cases above, complete the upsell purchase and verify that you are not shown any more upsell nudges.

**Scenario 4** 
* Verify status quo i.e current behavior is maintained for non-Premium plan purchases.
* Add a Personal plan to cart and complete signup. Verify that only the concierge upsell nudge is shown (no matter if you're in the `offer` or `noOffer` group for the `conciergeUpsellDial` test). 
* In browser console, type `localStorage.ABTests;` and verify that it does not list the `showBusinessPlanBump` test. 


**Scenario 5** _(Test that a failed purchase does not lead to a test assignment)_
* Go through the signup flow and checkout with a Premium plan. 
* In the Checkout page, click on PayPal as your payment method.
* After you're redirected to paypal.com, click the `Cancel and return to WordPress.com` link (if sandboxed, this will appear with different wording).
* In your browser console, type `localStorage.ABTests;`. The object should not include the `showBusinessPlanBump` test.
* Now try PayPal again and this time, complete the purchase
* You should still not be assigned to the test. Since we don't have a reliable way to know if a redirect payment type failed before determining the `successUrl`, we will exclude all redirect payment types from this test.

**Scenario 6** _(Test that a GiroPay purchase does not lead to a test assignment)_
* Go through signup flow while you are proxied through Europe. Pay for a Premium plan through Giropay and verify that you see the Concierge nudge. 
* Also verify that typing localstorage.ABTests in your browser console should not show the `showBusinessPlanBump` test in the list of assigned tests.

**Scenario 7** _(Test that that going Back doesn't show another offer)_
* When presented with the Business bump, click "Yes" but then when presented the Checkout click back in the browser and this time click "No" in the offer. You should be taken directly to Calypso.


**Test in the composite checkout**
* Where relevant, repeat all the cases listed above for the composite checkout flow - proxy through a US IP and assign yourself to `composite` variation of `showCompositeCheckout` test.
